### DISCUSSION TITLE
Is there a REST API for user CRUD in Superset?

### DISCUSSION THREAD

**Discussion by rindy8707** [apache/superset#23745] [Q&A / Help] [answered]
Is it possible to enable REST API for users in the URL(https://superset.apache.org/docs/api) even though I cannot find it currently?

**Comment by sebastianliebscher** [comment]
AFAIK it is not possible to CRUD users via Superset REST API.

You can either create users via CLI `flask fab create-user` or SQLAlchemy ORM [mentioned here](https://github.com/apache/superset/issues/10992).

Alternatively, you can use an OAuth Provider [mentioned in the docs](https://superset.apache.org/docs/installation/configuring-superset/#custom-oauth2-configuration).

**Comment by cwegener** [comment, marked as answer]
The underlying Security Management (including the above mentioned OAuth Provider!) in Superset is provided by Flask Application Builder (FAB).

FAB has all the REST API for user CRUD, but they are not enabled by default in Superset.

To enable this (beta) feature in Flask AppBuilder, you need to set the following configuration setting in your Superset configuration (Config setting. **NOT** Feature Flag):

```
FAB_ADD_SECURITY_API = True
```

Once configured, you can explore  the documentation in Swagger:

![image](https://user-images.githubusercontent.com/2950544/236130751-5aac1c8c-0a3b-472d-897b-765cd4d7cfc1.png)

**Reply by mdeshmu** [reply]
@sfirke should we document this in Superset docs?

**Reply by sfirke** [reply]
@mdeshmu great idea.  I think in the security section.  Would you like to send a PR?

**Reply by mdeshmu** [reply]
Yes

**Reply by PrateekKhatri** [reply]
Hi,

Any updates on this?
Also, when is the expected availability of this feature in the Helm charts?

**Reply by mdeshmu** [reply]
You can add it in your own helm overrides.

**Reply by PrateekKhatri** [reply]
Hi 
Can you please help me where to override this value.
Thanks in advance.

**Reply by mdeshmu** [reply]
https://github.com/apache/superset/blob/master/helm/superset/values.yaml#L133

**Reply by PrateekKhatri** [reply]
Hi,

Here is my values.yaml file 

`image:
  repository: apache/superset
  tag: "latest"
  pullPolicy: Always

configOverrides:
  FAB_ADD_SECURITY_API = True`


Getting below error:

`helm upgrade --install localsuperset --values ./charts/apache-superset-values.yaml superset/superset --version 0.10.10
Error: UPGRADE FAILED: template: superset/templates/secret-superset-config.yaml:33:8: executing "superset/templates/secret-superset-config.yaml" at <include "superset-config" .>: error calling include: template: superset/templates/_helpers.tpl:110:33: executing "superset-config" at <.Values.configOverrides>: range can't iterate over FAB_ADD_SECURITY_API = True`

**Reply by mdeshmu** [reply]
It should be a DICTIONARY of overrides

Please read the comments given in examples values.yaml carefully.

**Reply by PrateekKhatri** [reply]
Hi,
Thank you so much for your help. It's woking now.

**Reply by PrateekKhatri** [reply]
Hello,

I'm attempting to establish a connection to a ClickHouse database using an API. My schema for this connection is as follows:

```json
{
                    "allow_ctas": False,
                    "allow_cvas": False,
                    "allow_dml": False,
                    "allow_file_upload" : False,
                    "extra": "{\"allows_virtual_table_explore\":true}",
                    "sqlalchemy_uri": "clickhousedb+connect://<username>:<password>@clickhouse:8123/default",  
                    "cache_timeout": None,  
                    "expose_in_sqllab": True,  
                    "allow_run_async": True,
                    "allow_csv_upload": True, 
                    "database_name": "ClickHouse Connect", 
                    "parameters" : "DB",
                    "impersonate_user": False, 
                    "configuration_method" : "sqlalchemy_form",
                    "driver" : "connect",
                    "expose_in_sqllab" : True,
                    "database":  "default",
                    "uuid" : "clickhousdb"

 }
```

When I make a POST request to the API at `http://localhost:8088/api/v1/database/`, I encounter the following error:

"400 Bad Request: Issue 1011 - Superset encountered an unexpected error"

Is there something I might be missing or doing incorrectly?

**Reply by mdeshmu** [reply]
please use slack for quick responses w.r.t debugging

**Comment by jeffrey-atlastix** [comment]
For people who may want to do some role and user automations but don't want to turn security API on, you could use all these fab command lines. Specifically for creating roles in my case:
`superset fab import-roles -p ./role.json`
You can use `superset fab export-roles` to have a look at the expected role json format

**Reply by shivanshkaushikk** [reply]
hey @jeffrey-atlastix, any particular reason behind using fab commands instead of enabling security API? Wondering if there's any downside to using the security apis.

**Reply by mdeshmu** [reply]
probably because FF `FAB_ADD_SECURITY_API` is still in Beta. https://flask-appbuilder.readthedocs.io/en/latest/config.html

**Reply by tarungorayan** [reply]
> For people who may want to do some role and user automations but don't want to turn security API on, you could use all these fab command lines. Specifically for creating roles in my case: `superset fab import-roles -p ./role.json` You can use `superset fab export-roles` to have a look at the expected role json format

Can you explain this more? as we want to handle this process using our application so can we create a api using FAB or create script and api for that and use in application. Can we do something like this??
